### PR TITLE
Bump versions of maintenance-packages dependencies consumed in runtime

### PR DESCRIPTION
**Pull request by carlossanlop**
dotnet/runtime depends on these packages that we are now publishing from dotnet/maintenance-packages:

- Microsoft.Bcl.HashCode
- System.Buffers
- System.Memory
- System.Threading.Tasks.Extensions
- System.Runtime.CompilerServices.Unsafe

Bumping their versions to consume the new preview versions, available in the dotnet-libraries feed: https://dnceng.visualstudio.com/public/_artifacts/feed/dotnet-libraries

Related PRs:
- machinelearning: https://github.com/dotnet/runtime/pull/108806
- winforms: https://github.com/dotnet/winforms/pull/12313

### CODE REVIEW

File: eng/Versions.props [hunk 1]
```diff
@@ -129,9 +129,9 @@
     <SystemSecurityPrincipalWindowsVersion>5.0.0</SystemSecurityPrincipalWindowsVersion>
     <SystemSecurityPermissionsVersion>7.0.0</SystemSecurityPermissionsVersion>
     <SystemTextJsonVersion>10.0.0-alpha.1.24504.10</SystemTextJsonVersion>
-    <SystemRuntimeCompilerServicesUnsafeVersion>6.0.0</SystemRuntimeCompilerServicesUnsafeVersion>
+    <SystemRuntimeCompilerServicesUnsafeVersion>6.1.0-preview.1.24511.1</SystemRuntimeCompilerServicesUnsafeVersion>
     <SystemThreadingAccessControlVersion>7.0.0</SystemThreadingAccessControlVersion>
-    <SystemThreadingTasksExtensionsVersion>4.5.4</SystemThreadingTasksExtensionsVersion>
+    <SystemThreadingTasksExtensionsVersion>4.6.0-preview.1.24511.1</SystemThreadingTasksExtensionsVersion>
```
Comment (ericstj) [inline eng/Versions.props:134]: Looks to me like CLSCompliant attribute value on this assembly is causing errors.  @carlossanlop you double-check that the CLSCompliant attribute correct on all the m-p OOBs?

Comment (carlossanlop) [reply eng/Versions.props:134]: Yes.

Comment (carlossanlop) [reply eng/Versions.props:134]: Just for my own peace of mind, I'll post screenshot comparisons of them all. Give me some time to collect them.

**Comment by carlossanlop** [comment]
There's a CI failure indicating that a triple slash cref pointing to `MathF` is ambiguous. I see that the updated `System.Numerics.Vectors` OOB package defines `MathF` but it is an internal class there. I don't see how it could be conflicting.

**Comment by carlossanlop** [comment]
@ericstj @michaelgsharp @ViktorHofer  I don't think the problem is in the package. When comparing the DLLs with ILSpy, before and after maintenance-packages, they all contain the `MathF` class and it's _internal_. So I don't know what could be causing the cref conflict.

The old nupkg:
- net462
image
- netstandard2.0
image

The new preview nupkg:
- net462:
image
- netstandard2.0:
image

**Comment by ericstj** [comment]
The old package contained a reference assembly, which didn't include the internal `MathF` type.  That's why the new package is hitting this.  The new package only exposes the implementation, which has the internal type.  This is issue https://github.com/dotnet/roslyn/issues/71442.

We can workaround it by aliasing the problematic assembly, and then referring to it by the alias.  I'll submit a fix.

**Comment by carlossanlop** [comment]
Thanks for the fix @ericstj

>  I doubt customers will hit this bug

So there's no way a customer will do what Tensors is doing? Is Tensors special and unique enough because of the way runtime works and consumes the package?

**Comment by carlossanlop** [comment]
@MichaelSimons what specific thing abouy these nupkgs is considered a prebuilt?
We are hitting this error:

```
error : 6 new pre-builts discovered! 
Detailed usage report can be found at /__w/1/s/artifacts/sb/prebuilt-report/baseline-comparison.xml. 
See https://aka.ms/dotnet/prebuilts for guidance on what pre-builts are and how to eliminate them. 
Package IDs are: 
Microsoft.Bcl.HashCode.6.0.0-preview.1.24517.1 
System.Buffers.4.6.0-preview.1.24517.1 
System.Memory.4.6.0-preview.1.24517.1 
System.Numerics.Vectors.4.6.0-preview.1.24517.1 
System.Runtime.CompilerServices.Unsafe.6.1.0-preview.1.24517.1 
System.Threading.Tasks.Extensions.4.6.0-preview.1.24517.1
```

**Comment by ericstj** [comment]
> So there's no way a customer will do what Tensors is doing?

It's not impossible, just unlikely for a customer to `cref` `MathF` from their doc comments.

> what specific thing about these nupkgs is considered a prebuilt?

They're built from a repo that's not part of source build yet, and they aren't provided by SBRP.  Maybe we should bring maintenance-packages into source build @MichaelSimons.  In the meantime I think we can mark these as permitted prebuilts.

**Comment by MichaelSimons** [comment]
> They're built from a repo that's not part of source build yet, and they aren't provided by SBRP. Maybe we should bring maintenance-packages into source build @MichaelSimons. In the meantime I think we can mark these as permitted prebuilts.

If you do this the runtime flow will fail in the sdk repo until the dotnet/maintenance-packages repo is onboarded onto SB.  If you don't want to block the flow I would wait merging this PR until you have onboarded dotnet/maintenance-packages into source-build.

File: eng/Versions.props [hunk 2]
```diff
@@ -111,16 +111,17 @@
     <MicrosoftNETCoreILAsmVersion>10.0.0-alpha.1.24504.10</MicrosoftNETCoreILAsmVersion>
     <!-- Libraries dependencies -->
     <MicrosoftBclAsyncInterfacesVersion>6.0.0</MicrosoftBclAsyncInterfacesVersion>
-    <MicrosoftBclHashCodeVersion>1.1.1</MicrosoftBclHashCodeVersion>
+    <MicrosoftBclHashCodeVersion>6.0.0-preview.1.24517.1</MicrosoftBclHashCodeVersion>
```
Comment (carlossanlop) [inline eng/Versions.props:114]:
```suggestion
    <MicrosoftBclHashCodeVersion>6.0.0-preview.1.24529.4</MicrosoftBclHashCodeVersion>
```

File: eng/Versions.props [hunk 3]
```diff
@@ -111,16 +111,17 @@
...
     <MicrosoftWin32RegistryVersion>5.0.0</MicrosoftWin32RegistryVersion>
     <StyleCopAnalyzersVersion>1.2.0-beta.507</StyleCopAnalyzersVersion>
-    <SystemBuffersVersion>4.5.1</SystemBuffersVersion>
+    <SystemBuffersVersion>4.6.0-preview.1.24517.1</SystemBuffersVersion>
```
Comment (carlossanlop) [inline eng/Versions.props:117]:
```suggestion
    <SystemBuffersVersion>4.6.0-preview.1.24529.4</SystemBuffersVersion>
```

File: eng/Versions.props [hunk 4]
```diff
@@ -111,16 +111,17 @@
...
     <SystemComponentModelAnnotationsVersion>5.0.0</SystemComponentModelAnnotationsVersion>
     <SystemDataSqlClientVersion>4.8.6</SystemDataSqlClientVersion>
     <SystemDrawingCommonVersion>8.0.0</SystemDrawingCommonVersion>
     <SystemFormatsAsn1Version>8.0.1</SystemFormatsAsn1Version>
     <SystemIOFileSystemAccessControlVersion>5.0.0</SystemIOFileSystemAccessControlVersion>
-    <SystemMemoryVersion>4.5.5</SystemMemoryVersion>
+    <SystemMemoryVersion>4.6.0-preview.1.24517.1</SystemMemoryVersion>
```
Comment (carlossanlop) [inline eng/Versions.props:123]:
```suggestion
    <SystemMemoryVersion>4.6.0-preview.1.24529.4</SystemMemoryVersion>
```

File: eng/Versions.props [hunk 5]
```diff
@@ -129,9 +130,9 @@
     <SystemSecurityPrincipalWindowsVersion>5.0.0</SystemSecurityPrincipalWindowsVersion>
     <SystemSecurityPermissionsVersion>7.0.0</SystemSecurityPermissionsVersion>
     <SystemTextJsonVersion>10.0.0-alpha.1.24504.10</SystemTextJsonVersion>
-    <SystemRuntimeCompilerServicesUnsafeVersion>6.0.0</SystemRuntimeCompilerServicesUnsafeVersion>
+    <SystemRuntimeCompilerServicesUnsafeVersion>6.1.0-preview.1.24517.1</SystemRuntimeCompilerServicesUnsafeVersion>
```
Comment (carlossanlop) [inline eng/Versions.props:133]:
```suggestion
    <SystemRuntimeCompilerServicesUnsafeVersion>6.1.0-preview.1.24529.4</SystemRuntimeCompilerServicesUnsafeVersion>
```

File: eng/Versions.props [hunk 6]
```diff
@@ -129,9 +130,9 @@
...
     <SystemThreadingAccessControlVersion>7.0.0</SystemThreadingAccessControlVersion>
-    <SystemThreadingTasksExtensionsVersion>4.5.4</SystemThreadingTasksExtensionsVersion>
+    <SystemThreadingTasksExtensionsVersion>4.6.0-preview.1.24517.1</SystemThreadingTasksExtensionsVersion>
```
Comment (carlossanlop) [inline eng/Versions.props:135]:
```suggestion
    <SystemThreadingTasksExtensionsVersion>4.6.0-preview.1.24529.4</SystemThreadingTasksExtensionsVersion>
```

**Review by carlossanlop — COMMENTED** [review]
Bump the versions to the preview used in SBRP.

**Comment by carlossanlop** [comment]
I'll address the conflict after I merge https://github.com/dotnet/runtime/pull/109390 which unblocks this PR (so that I don't restart the CI unnecessarily).

**Comment by carlossanlop** [comment]
I triggered a new update from SBRP to bump the Numerics.Vectors package version to the correct one: https://github.com/dotnet/runtime/pull/109561

**Comment by carlossanlop** [comment]
@ViktorHofer I updated this PR to also consume the new System.Data.SqlClient prerelease version and there's an ApiCompat failure that makes sense: The old package used to contain a netcoreapp2.1 folder, but the new package contains a net6.0 and a net8.0 folder.

Is this failure expected and supressable, or should we add a placeholder for netcoreapp2.1 in System.Data.SqlClient?

```
src/libraries/apicompat/ApiCompat.proj(93,5): error MSB4018: (NETCORE_ENGINEERING_TELEMETRY=Build) The "Microsoft.DotNet.ApiCompat.Task.ValidateAssembliesTask" task failed unexpectedly.
System.IO.DirectoryNotFoundException: Could not find a part of the path '/__w/1/s/.packages/system.data.sqlclient/4.9.0-rtm.24561.10/lib/netcoreapp2.1/System.Data.SqlClient.dll'.
```

**Comment by carlossanlop** [comment]
@cheenamalhotra @David-Engel we're updating the SqlClient dependency in runtime to the latest. I suppose the warning is expected since it's now marked as obsolete. Should we suppress it or what do you suggest?

**Comment by David-Engel** [comment]
> we're updating the SqlClient dependency in runtime to the latest. I suppose the warning is expected since it's now marked as obsolete. Should we suppress it or what do you suggest?

Suppress, IMO.

**Comment by cheenamalhotra** [comment]
> @cheenamalhotra @David-Engel we're updating the SqlClient dependency in runtime to the latest. I suppose the warning is expected since it's now marked as obsolete. Should we suppress it or what do you suggest?

Where is it used? Can you create an issue to track migration to Microsoft.Data.SqlClient in future releases of .NET runtime?

File: src/libraries/System.Data.Common/tests/System.Data.Common.Tests.csproj [hunk 7]
```diff
@@ -1,6 +1,7 @@
 ﻿<Project Sdk="Microsoft.NET.Sdk">
   <PropertyGroup>
-    <NoWarn>$(NoWarn),0168,0169,0414,0219,0649</NoWarn>
+    <!-- 0618 suppresses warning to upgrade to Microsoft.Data.SqlClient. Remove from NoWarn list after upgrading. -->
+    <NoWarn>$(NoWarn),0168,0169,0414,0219,0649,0618</NoWarn>
```
[the remark this answers is not in the document]
Comment (carlossanlop) [reply src/libraries/System.Data.Common/tests/System.Data.Common.Tests.csproj:4]: Oh, good point. Alright, changing that.

**Comment by carlossanlop** [comment]
Sigh. Source build once again complaining about prebuilts:
```
.packages/microsoft.dotnet.arcade.sdk/10.0.0-beta.24556.2/tools/SourceBuild/AfterSourceBuild.proj(81,5): error : (NETCORE_ENGINEERING_TELEMETRY=AfterSourceBuild) 6 new pre-builts discovered! Detailed usage report can be found at /__w/1/s/artifacts/sb/prebuilt-report/baseline-comparison.xml.
See https://aka.ms/dotnet/prebuilts for guidance on what pre-builts are and how to eliminate them.
Package IDs are:
Microsoft.Bcl.HashCode.6.0.0
System.Buffers.4.6.0
System.Memory.4.6.0
System.Numerics.Vectors.4.6.0
System.Runtime.CompilerServices.Unsafe.6.1.0
System.Threading.Tasks.Extensions.4.6.0
```

**Comment by carlossanlop** [comment]
The wasm failures are unrelated:
- https://github.com/dotnet/runtime/issues/104349
- A build process died because the machine got reset

**Comment by MichaelSimons** [comment]
> Sigh. Source build once again complaining about prebuilts:

It looks like these need to be added to SBRP because of the update to use stable versions in https://github.com/dotnet/runtime/pull/108806/commits/ec8185fa13d240dc36ecbff24d903d638d91368e.  This was the plan discussed in https://github.com/dotnet/maintenance-packages/pull/140#issuecomment-2444383987.

**Comment by carlossanlop** [comment]
> It looks like these need to be added to SBRP

Here you go https://github.com/dotnet/source-build-reference-packages/pull/1082

**Comment by carlossanlop** [comment]
Darc flow PR https://github.com/dotnet/runtime/pull/110071

**Review by akoeplinger — APPROVED** [review]

**Review by ViktorHofer — APPROVED** [review]

**Pull request merged**